[One Workflow] Allow multiple workflow system actions

talboren opened the pull request:
### Summary
This PR refactors the system action framework to use a configurable allowMultipleSystemActions setting instead of hardcoded action type checks, as suggested in the code review feedback.

1. Added allowMultipleSystemActions property to ActionType interface
2. Updated Workflows connector to allow multiple instances
3. Refactored validation logic
4. Updated client-side UI logic

**Detection Rule**

<img width="827" height="670" alt="CleanShot 2025-11-26 at 11 35 02" src="https://github.com/user-attachments/assets/db5f6acd-9a8b-47bb-b8b7-77768146ddc7" />

**O11y Alert Rule**

<img width="1497" height="818" alt="CleanShot 2025-11-26 at 11 35 45" src="https://github.com/user-attachments/assets/1fca7767-c2b1-4510-93db-c45d75d4d0bb" />

talboren committed:
Allow multiple workflow system actions

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/365611)
* Commit: 0b8a528c29bf968526174f32f015241b27ee6f18

### Failed CI Steps
* [Jest Tests #20](https://buildkite.com/elastic/kibana-pull-request/builds/365611#019abfba-b923-4622-a977-96b49e6cee9c)
* [Jest Tests #20](https://buildkite.com/elastic/kibana-pull-request/builds/365611#019abfca-af17-4fc9-b52f-223f49bef494)
* [Scout Test Run Builder](https://buildkite.com/elastic/kibana-pull-request/builds/365611#019abf9c-7f3c-472a-9d0d-db8e3b35867f)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/365611#019abf9c-7f3c-472a-9d0d-db8e3b35867f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/365611/jobs/019abf9c-7f3c-472a-9d0d-db8e3b35867f/artifacts/019abfbf-4d00-4e0d-af09-a6173d06dad4) Scout Test Run Builder / fieldText
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/365611#019abf9c-7f3c-472a-9d0d-db8e3b35867f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/365611/jobs/019abf9c-7f3c-472a-9d0d-db8e3b35867f/artifacts/019abfbf-4d03-4225-83b3-690bef434ec3) Scout Test Run Builder / serverless-security - EUI testing wrapper: EuiFieldText - fieldText

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365611","number":365611,"commit":"0b8a528c29bf968526174f32f015241b27ee6f18"}],"number":365611}
buildkite-pr-comment-->

talboren committed:
Merge branch 'main' into multiple-workflow-system-action

talboren committed:
Merge branch 'main' into multiple-workflow-system-action

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/366922)
* Commit: 5b61b8ead49d708c02aec58f8674ff855164a1ec

### Failed CI Steps
* [Jest Tests #15](https://buildkite.com/elastic/kibana-pull-request/builds/366922#019ad5ab-71b0-4821-824a-a057a9cf285d)
* [Jest Tests #15](https://buildkite.com/elastic/kibana-pull-request/builds/366922#019ad5bb-1f58-4567-96c9-508cc4ec478b)
* [FTR Configs #107](https://buildkite.com/elastic/kibana-pull-request/builds/366922#019ad5ab-722d-4daf-a91f-f57836e57b19)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/366922#019ad5ab-722d-4daf-a91f-f57836e57b19) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/366922/jobs/019ad5ab-722d-4daf-a91f-f57836e57b19/artifacts/019ad5c8-a769-48ad-969c-4738f13ff4db) FTR Configs #<span></span>107 / discover - group 1 Discover CSV Export Generate CSV: new search generate a report using ES|QL

### History
* :broken_heart: [Build #365888](https://buildkite.com/elastic/kibana-pull-request/builds/365888) failed 5e0905fe5fdb3e761a9a8aaf0b57ca0ed4bd6c38
* :broken_heart: [Build #365611](https://buildkite.com/elastic/kibana-pull-request/builds/365611) failed 0b8a528c29bf968526174f32f015241b27ee6f18

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/366922","number":366922,"commit":"5b61b8ead49d708c02aec58f8674ff855164a1ec"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365888","number":365888,"commit":"5e0905fe5fdb3e761a9a8aaf0b57ca0ed4bd6c38"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365611","number":365611,"commit":"0b8a528c29bf968526174f32f015241b27ee6f18"}],"number":366922}
buildkite-pr-comment-->

talboren committed:
Merge branch 'main' into multiple-workflow-system-action

talboren committed:
Make system action multiple instances setting generic

This commit refactors the system actions framework to make it more
generic and extensible, addressing PR feedback to remove hardcoded
action type checks.

Changes:

1. Added allowMultipleSystemActions setting to ActionType
   - Introduced new optional field in ActionType interface that allows
     action types to declare whether they support multiple instances
     in a single rule
   - Updated both server (types.ts) and common (common/types.ts)
     ActionType interfaces

2. Updated all type definitions and schemas
   - Added field to ConnectorType schema and interface
   - Updated API response schemas (v1 and v2) to expose the new field
   - Modified transform functions to include the field in API responses
   - Updated ActionTypeRegistry.list() to return the field

3. Removed hardcoded action type checks
   - Previously: Had hardcoded Set with '.workflows' in validation logic
   - Now: Dynamically checks action type's allowMultipleSystemActions
     property by querying the action type registry
   - Only performs expensive checks when duplicates are detected
   - Maintains fast path for rules without duplicate system actions

4. Updated workflows connector
   - Added allowMultipleSystemActions: true to workflows connector type
   - This explicitly declares that workflows can be used multiple times
     per rule

5. Added comprehensive unit tests
   - Updated existing tests to mock the new field
   - Added test case for allowing multiple instances when the setting
     is enabled

Benefits:

- Generic and extensible: Any action type can opt-in without
  framework changes
- No hardcoded IDs: Framework doesn't need to know about specific
  action types
- Declarative: Action types declare their capabilities in their
  type definition
- Backward compatible: Existing action types default to false
  (previous behavior)
- Performance: Only validates when duplicates exist

How it works:

When a rule has multiple instances of the same system action:
1. Quick check: Compare Set size vs array length
2. If no duplicates: Skip validation (fast path)
3. If duplicates: Check if action type allows multiple instances
4. Throw error if not allowed, continue if allowed

talboren committed:
Merge branch 'multiple-workflow-system-action' of https://github.com/talboren/kibana into multiple-workflow-system-action

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

talboren committed:
ui to use allowMultipleSystemActions too

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

talboren committed:
fix tests

darnautov commented on x-pack/platform/plugins/shared/actions/common/routes/connector/response/schemas/v1.ts:
```diff
@@ -104,6 +104,14 @@ export const connectorTypeResponseSchema = schema.object({
+          'Indicates whether multiple instances of the same system action connector can be used in a single rule.',
+      },
+    })
+  ),
```
You only need to update the v2 (latest)
```suggestion
```

talboren committed:
fix test

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

talboren committed:
try fix

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

talboren committed:
Merge branch 'main' into multiple-workflow-system-action

cnasikas reviewed (commented):
Code review only! The approach LGTM. As we discussed offline, let's wait for this [PR](https://github.com/elastic/kibana/pull/244415) to get merged first. Also could you please add some integration tests? Specifically you can:

1. Register a new system action type that uses the new configuration here `x-pack/platform/test/alerting_api_integration/common/plugins/alerts/server/action_types.ts`.
2. Add a test here `x-pack/platform/test/alerting_api_integration/spaces_only/tests/alerting/group1/create.ts` that tests that you can create a rule with two system actions of the same type. You can follow the examples in `describe("system actions"`.
3. Same as above add a test here `x-pack/platform/test/alerting_api_integration/spaces_only/tests/alerting/group2/update.ts`.
4. Same as above add a test here `x-pack/platform/test/alerting_api_integration/spaces_only/tests/alerting/group2/bulk_edit.ts`.

Thanks!!

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

talboren committed:
simplify validateAndAuthorizeSystemActions

talboren committed:
fix test

talboren committed:
integration tests

kibanamachine committed:
Changes from node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/maintenance_window   --include-path /api/agent_builder --update

kibanamachine committed:
Changes from make api-docs

florent-leborgne reviewed (approved):
Docs change LGTM

skynetigor reviewed (approved):
LGTM

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

cnasikas committed:
Allow multiple system actions in the new rule form

talboren committed:
Merge branch 'multiple-workflow-system-action' of https://github.com/talboren/kibana into multiple-workflow-system-action

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

cnasikas reviewed (approved):
LGMT! I tested and everything is working as expected.

talboren committed:
fix tests

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

talboren committed:
fix test

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

talboren committed:
tests

talboren committed:
Merge branch 'main' of https://github.com/elastic/kibana into multiple-workflow-system-action

talboren committed:
Merge branch 'main' into multiple-workflow-system-action

talboren committed:
Merge branch 'main' into multiple-workflow-system-action
